### PR TITLE
Update dynamic fields in mapping on master even if parsing fails for the rest of the doc

### LINKED ISSUE #9851: [mappings] partially parsed documents can cause mapping loss
When a document is parsed only halfway but parsing exits with a MapperParsingException, the fields that were parsed till then are still in the local DocumentMapper but never added to the cluster state. Once the nodes are restarted the mapping is gone. I wore a test for it here https://github.com/brwe/elasticsearch/commit/52cd27c5735f678f20fb89c5eea8c746536f5c25#diff-defbaaff93b959a2f9a93e7167f6f345R246

We can potentailly fix this by also update the mapping on MapperParsingExceptions (https://github.com/brwe/elasticsearch/commit/52cd27c5735f678f20fb89c5eea8c746536f5c25#diff-9669e07f0556311d187e534e321a0393R422) but we would probably need to check first if the mapper was actually changed, otherwise we might end with one update for each Exception.

### PR DESCRIPTION
**Pull request by brwe**
...ils for the rest of doc

The local DocumentMapper is updated while parsing and dynamic fields are added before
parsing has finished. If parsing fails after a dynamic field has been added already
then the field was not added to the cluster state but was present in the local mapper of this
node. New documents with the same field would not necessarily cause an update either and
after restarting the node the mapping for these fields were lost. Instead the new fields
should always be updated.

closes #9851

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 1]
```diff
@@ -494,7 +494,7 @@ public ParsedDocument parse(SourceToParse source, @Nullable ParseListener listen
         } catch (Throwable e) {
             // if its already a mapper parsing exception, no need to wrap it...
             if (e instanceof MapperParsingException) {
-                throw (MapperParsingException) e;
+                throw ((MapperParsingException) e).setMappingsModified(context.mappingsModified());
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: Could we do it in such a way that we do not need to modify the incoming exception in-place?

Comment (brwe) [inline src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: Not sure what you mean. I added it now in the constructor of the MapperParsingException. Is this what you meant?

Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: Yes! I like it much better!

**Comment by jpountz** [comment]
LGTM

**Comment by jpountz** [comment]
@brwe Since this change looks quite significant, I don't think it should be pushed to 1.4.5.

**Comment by brwe** [comment]
That was not a good idea - the change of MapperParsingException seemes to have screwed up bwc for Exception serialization, see for example http://build-us-00.elasticsearch.org/job/es_bwc_1x/8385/
I reverted the commit now.

**Comment by brwe** [comment]
Ok, new plan. Instead of changing the mapper parsing exception we actually can just get the context from the document mapper and ask if the mapping was changed - much easier and I cannot see any downside to it. @jpountz would you mind taking another look?

**Comment by jpountz** [comment]
I'm not too happy with getting back to the cache to check whether mappings have been modified, I liked the previous solution better. But on ther other hand, I don't have a better idea. Can we maybe just use this approach on 1.x and keep the previous approach on master?

**Comment by brwe** [comment]
Ok, I'll do that.

**Pull request closed without merging**